lib/ansible/plugins/inventory/linode.py

```diff
@@ -125,18 +132,85 @@ def _filter_by_config(self, regions, types):
                 if instance.type.id in types
             ]
 
-    def _add_instances_to_groups(self):
+    def _get_instance_public_ip(self, instance):
+        """Returns the public IP address of the instance"""
+        if instance.ipv4 and len(instance.ipv4) > 0:
+            public_ip = None
+            for ip in instance.ipv4:
+                if not ipaddress.ip_address(ip).is_private:
+                    public_ip = ip
+            if public_ip:
+                return public_ip
+            else:
+                raise AnsibleError(
+                    'Expected public IP not found'
+                )
+        else:
+            raise AnsibleError(
+                'Instance IPv4 address is missing or empty'
+            )
+
+    def _get_instance_private_ip(self, instance):
+        """Returns the private IP address of the instance"""
+        if instance.ipv4 and len(instance.ipv4) > 0:
+            private_ip = None
+            for ip in instance.ipv4:
+                if ipaddress.ip_address(ip).is_private:
+                    private_ip = ip
+            if private_ip:
+                return private_ip
+            else:
+                raise AnsibleError(
+                    'Expected public IP not found'
+                )
+        else:
+            raise AnsibleError(
+                'Instance IPv4 address is missing or empty'
+            )
+
+    def _add_instances_to_groups(self, instance_access):
         """Add instance names to their dynamic inventory groups."""
         for instance in self.instances:
-            self.inventory.add_host(instance.label, group=instance.group)
+            if instance_access == 'public_ip':
+                host_identifier = self._get_instance_public_ip(instance)
+            elif instance_access == 'private_ip':
+                host_identifier = self._get_instance_private_ip(instance)
+            elif instance_access == 'hostname':
+                host_identifier = instance.label
+            else:
+                raise AnsibleError(
+                    'Instance Access setting is not valid'
+                )
+            self.inventory.add_host(host_identifier, group=instance.group)
 
-    def _add_hostvars_for_instances(self):
+    def _add_hostvars_for_instances(self, instance_access):
         """Add hostvars for instances in the dynamic inventory."""
         for instance in self.instances:
+            if instance_access == 'public_ip':
+                host_identifier = self._get_instance_public_ip(instance)
+            elif instance_access == 'private_ip':
+                host_identifier = self._get_instance_private_ip(instance)
+            elif instance_access == 'hostname':
+                host_identifier = instance.label
+            else:
+                raise AnsibleError(
```

Comment:
If we use `choices` in the documentation fragment, we can avoid this (invalid values will never be allowed here).